### PR TITLE
Internally generate the model URL array when the provided URL for `loadModelFromUrl` method is from a single shard of a model split with the `gguf-split` tool

### LINKED ISSUE #58: Made a function to build the Model URL Array when detecting the url has the gguf-split pattern `-<number>-of-<number>.gguf`. Would it fit in the lib?
Here's the function:

```typescript
function parseModelUrl(url: string) {
  const urlPartsRegex = /(.*)-(\d{5})-of-(\d{5})\.gguf$/;

  const matches = url.match(urlPartsRegex);

  if (!matches || matches.length !== 4) return url;

  const baseURL = matches[1];

  const paddedShardsAmount = matches[3];

  const paddedShardNumbers = Array.from(
    { length: Number(paddedShardsAmount) },
    (_, i) => (i + 1).toString().padStart(5, "0"),
  );

  return paddedShardNumbers.map(
    (paddedShardNumber) =>
      `${baseURL}-${paddedShardNumber}-of-${paddedShardsAmount}.gguf`,
  );
}
```

## Output examples

When a split GGUF URL is detected, it returns the array:

```typescript
parseModelUrl("https://huggingface.co/Felladrin/gguf-sharded-Phi-3-mini-4k-instruct-iMat/resolve/main/phi-3-mini-4k-instruct-imat-Q5_K_M.shard-00001-of-00051.gguf");
// Outputs:
// [
//     "https://huggingface.co/Felladrin/gguf-sharded-Phi-3-mini-4k-instruct-iMat/resolve/main/phi-3-mini-4k-instruct-imat-Q5_K_M.shard-00001-of-00051.gguf",
//     "https://huggingface.co/Felladrin/gguf-sharded-Phi-3-mini-4k-instruct-iMat/resolve/main/phi-3-mini-4k-instruct-imat-Q5_K_M.shard-00002-of-00051.gguf",
//     (...)
//     "https://huggingface.co/Felladrin/gguf-sharded-Phi-3-mini-4k-instruct-iMat/resolve/main/phi-3-mini-4k-instruct-imat-Q5_K_M.shard-00050-of-00051.gguf",
//     "https://huggingface.co/Felladrin/gguf-sharded-Phi-3-mini-4k-instruct-iMat/resolve/main/phi-3-mini-4k-instruct-imat-Q5_K_M.shard-00051-of-00051.gguf"
// ]
```

Otherwise, it returns the URL unchanged:

```typescript
parseModelUrl("https://huggingface.co/Felladrin/gguf-Pythia-31M-Chat-v1/resolve/main/Pythia-31M-Chat-v1.Q8_0.gguf")
// Outputs:
// 'https://huggingface.co/Felladrin/gguf-Pythia-31M-Chat-v1/resolve/main/Pythia-31M-Chat-v1.Q8_0.gguf'
```

---

Do you think it could be useful for the lib, to also accept a single URL of a split gguf in `loadModelFromUrl`?

[If the user sets an array, it wouldn't need to be parsed, as it's understandable that the user wants to use a custom list of ggufs. But as most will use the gguf-split's default naming, it could save some time.]

### PR DESCRIPTION
**Pull request by felladrin**
Resolves #58

## Example

Loading a 7-shards model:

```typescript
import { Wllama } from './esm/index.js';

(async () => {
  const CONFIG_PATHS = {
    'single-thread/wllama.js'       : './esm/single-thread/wllama.js',
    'single-thread/wllama.wasm'     : './esm/single-thread/wllama.wasm',
    'multi-thread/wllama.js'        : './esm/multi-thread/wllama.js',
    'multi-thread/wllama.wasm'      : './esm/multi-thread/wllama.wasm',
    'multi-thread/wllama.worker.mjs': './esm/multi-thread/wllama.worker.mjs',
  };

  const wllama = new Wllama(CONFIG_PATHS);

  const progressCallback =  ({ loaded, total }) => {
    const progressPercentage = Math.round((loaded / total) * 100);
    console.log(`Downloading... ${progressPercentage}%`);
  };

  await wllama.loadModelFromUrl(
    "https://huggingface.co/Felladrin/gguf-sharded-Llama-160M-Chat-v1/resolve/main/Llama-160M-Chat-v1.Q8_0.shard-00001-of-00007.gguf",
    {
      progressCallback,
    }
  );

  const outputText = await wllama.createCompletion("The best way to become healthier is", {
    nPredict: 250,
    sampling: {
      temp: 0.5,
      top_k: 40,
      top_p: 0.9,
    },
  });

  console.log(outputText);
})();
```

### CODE REVIEW

File: src/wllama.ts [hunk 1]
```diff
@@ -206,6 +206,33 @@ export class Wllama {
     return this.useMultiThread;
   }
 
+  /**
+   * Parses a model URL and returns an array of URLs based on the following patterns:
+   * If the input URL is an array, it returns the array itself.
+   * If the input URL is a string in the `gguf-split` format, it returns an array containing the URL of each shard in ascending order.
+   * Otherwise, it returns an array containing the input URL as a single element array.
+   */
+  parseModelUrl(modelUrl: string | string[]) {
```
Comment (ngxson) [inline src/wllama.ts:215]:
```suggestion
  private parseModelUrl(modelUrl: string | string[]): string[] {
```

Also maybe add `@param` for the inline docs

File: src/wllama.ts [hunk 2]
```diff
@@ -206,6 +206,33 @@ export class Wllama {
...
+    if (Array.isArray(modelUrl)) return modelUrl;
+
```
Comment (ngxson) [inline src/wllama.ts:217]: Could we remove the empty lines?

File: src/wllama.ts [hunk 3]
```diff
@@ -206,6 +206,33 @@ export class Wllama {
...
+  
+    if (!matches || matches.length !== 4) return [modelUrl];
+  
+    const baseURL = matches[1];
+  
+    const paddedShardsAmount = matches[3];
+  
+    const paddedShardNumbers = Array.from(
+      { length: Number(paddedShardsAmount) },
+      (_, i) => (i + 1).toString().padStart(5, "0")
```
Comment (ngxson) [inline src/wllama.ts:230]:
```suggestion
      (_, i) => (i + 1).toString().padStart(5, '0')
```

File: src/wllama.ts [hunk 4]
```diff
@@ -206,6 +206,33 @@ export class Wllama {
...
+  /**
+   * Parses a model URL and returns an array of URLs based on the following patterns:
+   * If the input URL is an array, it returns the array itself.
+   * If the input URL is a string in the `gguf-split` format, it returns an array containing the URL of each shard in ascending order.
+   * Otherwise, it returns an array containing the input URL as a single element array.
+   */
+  parseModelUrl(modelUrl: string | string[]) {
+    if (Array.isArray(modelUrl)) return modelUrl;
+  
+    const urlPartsRegex = /(.*)-(\d{5})-of-(\d{5})\.gguf$/;
```
Comment (ngxson) [inline src/wllama.ts:218]: FYI, there is a feature called [Named capturing group](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Named_capturing_group) that we can use here:

```suggestion
    const urlPartsRegex = /(?<baseURL>.*)-(?<current>\d{5})-of-(?<total>\d{5})\.gguf$/;
```

**Review by ngxson — COMMENTED** [review]
LGTM overall, just some small comments about coding style

**Comment by ngxson** [comment]
Also could you update README example for `loadModelFromUrl`? Thanks!

**Comment by felladrin** [comment]
👌 Will update it today later!

**Comment by flatsiedatsie** [comment]
Looking forward to it, as it will make it much simpler to implement the "try an AI by sharing/clicking on a link" feature I described earlier (which I'm implementing now).

**Comment by felladrin** [comment]
Had some unforeseen circumstances that prevented me from doing this yesterday.

Gonna update it now.

**Comment by flatsiedatsie** [comment]
Cool. I've just about finished implementing the "try a model by clicking on a link" feature.

**Comment by felladrin** [comment]
Updated and ready for another review 🙌

**Comment by felladrin** [comment]
Ah, wait, one more commit, on the Readme.

**Comment by flatsiedatsie** [comment]
I just ran into this hickup on the 1.1 version, but I thought I'd share it here.

Screenshot 2024-06-05 at 00 13 20

**Comment by felladrin** [comment]
> Screenshot 2024-06-05 at 00 13 20

@flatsiedatsie, I confirmed it's an issue present on the main branch, and that it's not related to nor fixed by the changes on this PR, so I turned it in a new issue:
- https://github.com/ngxson/wllama/issues/63

**Review by ngxson — APPROVED** [review]
:+1: Looks good. Thank you!

**Pull request merged**